chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiser.java

```diff
@@ -233,121 +231,138 @@ private static int getDimIndex(String fieldName, String prefix) {
      */
-    public static DataSet readDataSetFromByteArray(final IoBuffer readBuffer) { // NOPMD
+    public DataSet readDataSetFromByteArray() { // NOPMD
         final DataSetBuilder builder = new DataSetBuilder();
-        final long initialPosition = readBuffer.position();
-        final HeaderInfo bufferHeader = BinarySerialiser.checkHeaderInfo(readBuffer);
-        LOGGER.atTrace().addArgument(bufferHeader).log("read header = {}");
-
-        readBuffer.position(initialPosition);
-        FieldHeader fieldRoot = BinarySerialiser.parseIoStream(readBuffer);
-        fieldRoot = fieldRoot.getChildren().get(0); // N.B. old convention did not have a ROOT object
+        FieldDescription root = ioSerialiser.parseIoStream(false);
+        final FieldDescription fieldRoot = root.getChildren().get(0);
         // parsed until end of buffer
 
-        parseHeaders(readBuffer, builder, fieldRoot.getChildren());
+        parseHeaders(ioSerialiser, builder, fieldRoot);
 
         if (isMetaDataSerialised()) {
-            parseMetaData(readBuffer, builder, fieldRoot.getChildren());
+            parseMetaData(ioSerialiser, builder, fieldRoot);
         }
 
         if (isDataLablesSerialised()) {
-            parseDataLabels(readBuffer, builder, fieldRoot.getChildren());
+            parseDataLabels(builder, fieldRoot);
         }
 
-        parseNumericData(readBuffer, builder, fieldRoot.getChildren());
+        parseNumericData(ioSerialiser, builder, fieldRoot);
 
+        if (root.getChildren().size() != 2) {
+            throw new IllegalArgumentException("fieldRoot children-count != 2: " + fieldRoot.getChildren().size());
+        }
+        final FieldDescription endMarker = root.getChildren().get(1);
+        if (endMarker.getDataType() != DataType.END_MARKER) {
+            throw new IllegalArgumentException("fieldRoot END_MARKER expected but found: " + endMarker);
+        }
+        // move read position to after end marker
+        ioSerialiser.getBuffer().position(endMarker.getDataStartPosition() + endMarker.getDataSize());
         return builder.build();
     }
 
-    public static void setDataLablesSerialised(final boolean state) {
+    public void setDataLablesSerialised(final boolean state) {
         transmitDataLabels = state;
     }
 
-    public static void setMetaDataSerialised(final boolean state) {
+    public void setMetaDataSerialised(final boolean state) {
         transmitMetaData = state;
     }
 
-    private static float[] toFloats(final double[] input) {
+    private float[] toFloats(final double[] input) {
         final float[] floatArray = new float[input.length];
         for (int i = 0; i < input.length; i++) {
             floatArray[i] = (float) input[i];
         }
         return floatArray;
     }
 
-    protected static void writeDataLabelsToStream(final IoBuffer buffer, final DataSet dataSet) {
+    protected void writeDataLabelsToStream(final DataSet dataSet) {
```

Comment:
Codacy found an issue: [The method 'writeDataLabelsToStream(DataSet)' has an NPath complexity of 320, current threshold is 200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5905529)